bindings/rust/s2n-tls/src/callbacks.rs

```diff
@@ -57,91 +57,180 @@ where
+/// when implementing an async callback, eg. [`ClientHelloCallback`], if it wants
+/// to run an asynchronous operation (disk read, network call). The application
+/// can return an error ([`Err(error::Error::application())`]), to indicate
+/// connection failure.
 ///
-/// Implementations should capture any arguments passed to the callback
-/// so that they will be available for every call to [`AsyncCallback::poll`].
-/// [`AsyncCallback::poll`] should handle retrieving the callback from
-/// the connection or config, passing it the stored arguments,
-/// and calling any "mark done" style methods necessary to unblock the
-/// connection once the callback has succeeded.
-pub(crate) trait AsyncCallback {
-    fn poll(&mut self, conn: &mut Connection) -> Poll<Result<(), Error>>;
+/// [`ConfigResolver`] should be used if the application wants to set a new
+/// [`Config`] on the connection.
+pub trait ConnectionFuture {
+    fn poll(
+        self: Pin<&mut Self>,
+        connection: &mut Connection,
+        ctx: &mut core::task::Context,
+    ) -> Poll<Result<(), Error>>;
+}
+
+pin_project! {
+/// An implementation of [`ConnectionFuture`] which resolves the provided
+/// future and sets the config on the [`connection::Connection`].
+pub struct ConfigResolver<F: Future<Output = Result<Config, Error>>> {
+    #[pin]
+    fut: F,
+}
+}
+
+impl<F: Future<Output = Result<Config, Error>>> ConfigResolver<F> {
+    pub fn new(fut: F) -> Self {
+        ConfigResolver { fut }
+    }
+}
+
+impl<F: Future<Output = Result<Config, Error>>> ConnectionFuture for ConfigResolver<F> {
+    fn poll(
+        self: Pin<&mut Self>,
+        connection: &mut Connection,
+        ctx: &mut core::task::Context,
+    ) -> Poll<Result<(), Error>> {
+        let this = self.project();
+        let config = match this.fut.poll(ctx) {
+            Poll::Ready(config) => config?,
+            Poll::Pending => return Poll::Pending,
+        };
+
+        connection.set_config(config)?;
+
+        Poll::Ready(Ok(()))
+    }
 }
 
 /// A trait for the callback executed after parsing the TLS Client Hello.
 ///
 /// Use in conjunction with
 /// [config::Builder::set_client_hello_callback](`crate::config::Builder::set_client_hello_callback()`).
 pub trait ClientHelloCallback {
-    fn poll_client_hello(&self, connection: &mut Connection) -> Poll<Result<(), Error>>;
+    /// The application can return a `Ok(None)` to resolve the client_hello_callback
+    /// synchronously or return a `Ok(Some(ConnectionFuture))` if it wants to
+    /// run some asynchronous task beforing resolving the callback.
+    ///
+    /// [`ConfigResolver`], which implements [`ConnectionFuture`] can be
+    /// returned if the application wants to set a new [`Config`] on the connection.
+    ///
+    /// If the server_name is used to configure the connection then the application
+    /// should call [`connection::Connection::server_name_extension_used()`].
```

Comment:
what happens if this isn't done? Maybe is should say `must` instead of `should`

Reply:
From what I understand there is nothing forcing the application from calling this and the connection still succeeds if the application fails to call this. I tried finding what this is used for and what happens if application fail to call it and didnt find much. Its a frustrating RFC requirement and I dont understand why it exits.

https://github.com/aws/s2n-tls/blob/99b43512ce6ec9e3a1b848d9b042a0fe7f7daefe/tls/s2n_connection.h#L86-L90